### PR TITLE
Offload remote inbox notifications engine run using action-scheduler

### LINKED ISSUE #6189: Remote Inbox: Performance Tweaks
**Describe the bug**
Related: #6181

When looking into the Remote Inbox bug ( see related issue above ) I noticed that [the hook in `StoredStateSetupForProducts`](https://github.com/woocommerce/woocommerce-admin/blob/main/src/RemoteInboxNotifications/StoredStateSetupForProducts.php#L22) is I believe inadvertently calling `RemoteInboxNotificationsEngine::run();` anytime an already published product is being updated.

I think the intention of this hook is to only run when a product is transitioning from draft -> published but cc @becdetat  for clarification.

__Suggested Actions__
- [ ] ~~Add stronger checks to the `run_on_transition_post_status` method to ensure the logic only runs when a product transitions to publish~~
- [ ] ~~Investigate if we could maybe use some woo specific hooks here instead, maybe `woocommerce_new_product`, I feel like Bec looked into these but they didn't work as expected.~~
- [ ] I would also like us to consider not triggering a download of remote inbox specs via `RemoteInboxNotificationsEngine::run()` could we instead run the engine using persisted note rules if they exist? Sites _should_ have a cached copy of remote inbox specs from the prior day's cron job, so seems like this should be okay.
- [ ] If a full run of the remote inbox is required, I think we should consider placing the execution of that code path in an actionscheduler job so we aren't doing a wc_remote_get during the product update request.

cc @ilyasfoo @moon0326 for any further thoughts since you both spent some time on this.

### PR DESCRIPTION
**Pull request by becdetat**
Partially fixes #6189

This uses action-scheduler to offload running the remote inbox notifications engine run when adding or importing products.

### Detailed test instructions:

1. delete all products
5. delete all admin notes (using the test helper plugin)
2. add test feed url to `DataSourcePoller`: `https://gist.githubusercontent.com/becdetat/2d9895ad8e159ece76e8b3d029b09e8b/raw/30ad24cef0945220fbe351fa5af5d5cca8f4cf55/three-products.json`
6. add three new products
7. wait a minute for action-scheduler to do it's magic
8. visit WC Admin's home page
9. The new test note should be present

image

### CODE REVIEW

**Review by ilyasfoo — COMMENTED** [review]
When I tested this, I hit a couple of snags. Not all of them are related to the PR's changes but they do interfere with testing:
- The note did not come out, and I think it was because 
`wc_remote_inbox_notifications_specs` option already exists and had value from the original datasource url, thus the poller did not fetch new. I had to delete the option.
- `wc_remote_inbox_notifications_stored_state` has `new_product_count` set to 1, and according to the [line here](https://github.com/woocommerce/woocommerce-admin/blob/6acebc08cb7be11bce52801a94458ed302096c07/src/RemoteInboxNotifications/ProductCountRuleProcessor.php#L47), this was calculated to 4 instead of 3. I do have old products trashed, but not permanently deleted.
- When adding the third product, it seems that `as_enqueue_async_action` did not result in running the specs. I confirm that this [line runs](https://github.com/woocommerce/woocommerce-admin/blob/6acebc08cb7be11bce52801a94458ed302096c07/src/RemoteInboxNotifications/StoredStateSetupForProducts.php#L133). I've tried to wait for more than 5 minutes, navigated to other pages in wp-admin, but they did not trigger a spec run.

I'm not sure if I'm doing anything wrong, maybe if you could enlighten me @becdetat

**Comment by becdetat** [comment]
Thanks @ilyasfoo 

I was hooking up the action-scheduler handler in the `admin_init` hook, which doesn't get executed on a WP Cron task, so the handler never got called most (perhaps 95%) of the time. It happened to work (when it didn't execute using WP Cron) when I was testing, so I thought it was ok before pushing this PR.

I've changed where the handler is hooked up (in the `init` hook with a priority of 0) and removed the product count workaround as you noticed (not required now since the product count processor happens in a different request to where the product is actually added).

File: src/RemoteInboxNotifications/ProductCountRuleProcessor.php [hunk 1]
```diff
@@ -40,14 +40,10 @@ public function __construct( $product_query = null ) {
 	 * @return bool The result of the operation.
 	 */
 	public function process( $rule, $stored_state ) {
-		$products          = $this->product_query->get_products();
-		$new_product_count = property_exists( $stored_state, 'new_product_count' )
-			? $stored_state->new_product_count
-			: 0;
-		$count             = $products->total + $new_product_count;
+		$products = $this->product_query->get_products();
```
Comment (ilyasfoo) [inline src/RemoteInboxNotifications/ProductCountRuleProcessor.php:43]: Nice, thanks for making this fix! I thought it was a bit confusing. I was going to say to add a test for this change, but then I saw the existing tests and I think it's a simple enough logic to not require more tests.

File: src/RemoteInboxNotifications/RemoteInboxNotificationsEngine.php [hunk 2]
```diff
@@ -73,12 +76,21 @@ public static function on_admin_init() {
 
 		add_action( 'activated_plugin', array( __CLASS__, 'run' ) );
```
Comment (ilyasfoo) [inline src/RemoteInboxNotifications/RemoteInboxNotificationsEngine.php:77]: I wonder if it's worth considering making the engine run here async as well?

If this is agreed, I would also suggest adding the `woocommerce_activated_plugin` action to cater for [wcadmin embedded](https://github.com/woocommerce/woocommerce/pull/28698).

**Review by ilyasfoo — COMMENTED** [review]
@becdetat I've tested this twice in a row and confirmed the async action works reliably both times. This is shaping up nicely! 

I have a suggestion to add async to the initial run when plugin activation as well, would love to know your thoughts, as I couldn't think of any case against it right now.

File: src/RemoteInboxNotifications/RemoteInboxNotificationsEngine.php (hunk 2 above)
Comment (becdetat) [reply src/RemoteInboxNotifications/RemoteInboxNotificationsEngine.php:77]: This is probably a good idea, I can't see a reason not to make it async apart from introducing complexity when there isn't a huge amount of gain - plugins would generally be activated much less than products are added/edited, and there isn't much much overhead in running the engine as it doesn't actually pull down from the feeds except on a first run when WCA is installed.

The extra action makes sense too. I'll create these as a follow-up ticket though.

Comment (ilyasfoo) [reply src/RemoteInboxNotifications/RemoteInboxNotificationsEngine.php:77]: I agree, I think the added complexity might not worth the performance add. Thanks for making the follow-up ticket for the added action!

**Review by ilyasfoo — APPROVED** [review]
Really great work here, thanks! 🚢

**Pull request merged**